pkg/R/utils.R

```diff
@@ -16,51 +16,41 @@ convertJListToRList <- function(jList, flatten, logicalUpperBound = NULL, serial
   results <- if (arrSize > 0) {
     lapply(0:(arrSize - 1),
            function(index) {
-             jElem <- .jcall(jList,
-                             "Ljava/lang/Object;",
-                             "get",
-                             as.integer(index))
+             obj <- callJMethod(jList, "get", as.integer(index))
 
              # Assume it is either an R object or a Java obj ref.
-             obj <- .jsimplify(jElem)
-
-             if (inherits(obj, "jobjRef") && .jinstanceof(obj, "[B")) {
-               # RDD[Array[Byte]]. `obj` is a whole partition.
-
-               rRaw <- .jevalArray(.jcastToArray(jElem))
-               res <- unserialize(rRaw)
-
-               # For serialized datasets, `obj` (and `rRaw`) here corresponds to
-               # one whole partition dense-packed together. We deserialize the
-               # whole partition first, then cap the number of elements to be returned.
-
-               # TODO: is it possible to distinguish element boundary so that we can
-               # unserialize only what we need?
-               if (!is.null(logicalUpperBound)) {
+             if (inherits(obj, "jobj")) {
+               if (grepl("scala.Tuple2", obj$id)) {
```

Comment:
Hmm thats a good point. Actually I am not sure we need a new RPC. Can't we do this dynamically ? For example

```
isInstance <- function(jobj) {
  cls <- SparkR:::callJMethod(jobj, "getClass")
  SparkR:::callJMethod(cls, "isInstance", jobj)
}
```